### DISCUSSION TITLE
Make ByteBuff more efficient: Use of Unsafe.putXXX, Unsafe.getXXX maybe more efficient

### DISCUSSION THREAD

**Discussion by sirinath** [activej/activej#180] [Ideas]
In `ByteBuf` use of `Unsafe` maybe more efficient than what is [used now](https://github.com/activej/activej/blob/v3.0.1/core-bytebuf/src/main/java/io/activej/bytebuf/ByteBuf.java#L985). [This is another library](https://github.com/real-logic/agrona/blob/master/agrona/src/main/java/org/agrona/concurrent/UnsafeBuffer.java) using the `Unsafe` approach. Also one can use `MethodHandles::byteArrayViewVarHandle`.

Ref:
- https://github.com/activej/activej/blob/v3.0.1/core-bytebuf/src/main/java/io/activej/bytebuf/ByteBuf.java#L985
- https://github.com/real-logic/agrona/blob/master/agrona/src/main/java/org/agrona/concurrent/UnsafeBuffer.java
- https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/MethodHandles.html#byteArrayViewVarHandle-java.lang.Class-java.nio.ByteOrder-

**Comment by eduard-vasinskyi** [comment]
Hello, @sirinath 
Even if we use `Unsafe` for `ByteBuf`’s `putXXX`, `getXXX` operations, we would still have to do bounds check as well as moving head/tail to the new position. As a result, we wouldn’t get a performance boost.
`ByteBuf` is intended to work with bulk operations that operate on byte arrays. If you need to work with bytes individually, you can use `ByteBuf::array` directly.
